### PR TITLE
[PLUGIN-1732] - Successfactors Proxy Implementation

### PR DESCRIPTION
**Pull request by Shubhangi-cs**
Proxy Configuration : 

Proxy URL: Users can enter the URL of the proxy server. This URL should contain the protocol (e.g., http or https), the proxy server's address, and the port number (e.g.,[ http://proxy.example.com:8080](http://proxy.example.com:8080/)).

Username (Optional): If the proxy server requires authentication, users can provide the username.

Password (Optional): If the proxy server requires authentication, users can supply the password.

This enhancement will enable users to set up and use Proxy settings in the Successfactors plugin, allowing them to route their requests through a proxy server when communicating with external endpoints.

https://cdap.atlassian.net/browse/PLUGIN-1732

### CODE REVIEW

File: src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java [hunk 1]
```diff
@@ -189,6 +197,39 @@ private Request buildRequest(URL endpoint, String mediaType) {
...
+  private OkHttpClient buildConfiguredClient(String proxyUrl, String proxyUsername, String proxyPassword)
+    throws MalformedURLException, TransportException {
+    OkHttpClient.Builder builder = getConfiguredClient();
+
+    if (proxyUrl != null && !proxyUrl.isEmpty()) {
+      URL url = new URL(proxyUrl);
+      Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(url.getHost(), url.getPort()));
+      builder.proxy(proxy);
+
+      if (SuccessFactorsUtil.isNotNullOrEmpty(proxyUsername) && SuccessFactorsUtil.isNotNullOrEmpty(proxyPassword)) {
```
Comment (sahusanket) [inline src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java:217]: use 
`!Strings.isNullOrEmpty(proxyUrl)` 
from 
`import com.google.common.base.Strings;`

File: src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java [hunk 2]
```diff
@@ -50,15 +61,11 @@ public class SuccessFactorsTransporter {
   private static final long CONNECTION_TIMEOUT = 300;
   private static final long WAIT_TIME = 5;
   private static final long MAX_NUMBER_OF_RETRY_ATTEMPTS = 5;
-
-  private final String username;
-  private final String password;
+  private SuccessFactorsConnectorConfig config;
```
Comment (sahusanket) [inline src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java:64]: NIT : add empty line after private static final variables

File: src/main/java/io/cdap/plugin/successfactors/source/SuccessFactorsSource.java [hunk 3]
```diff
@@ -136,7 +136,16 @@ private Schema getOutputSchema(FailureCollector failureCollector) {
       } catch (TransportException te) {
         String errorMsg = ExceptionParser.buildTransportError(te);
         errorMsg = ResourceConstants.ERR_ODATA_SERVICE_CALL.getMsgForKeyWithCode(errorMsg);
-        failureCollector.addFailure(errorMsg, null).withConfigProperty(SuccessFactorsPluginConfig.BASE_URL);
+        if ((errorMsg.contains("proxy") || errorMsg.contains("Connection reset"))) {
```
Comment (sahusanket) [inline src/main/java/io/cdap/plugin/successfactors/source/SuccessFactorsSource.java]: Is there a documentation that maps these KEY words to the given properties ? 

`Connection reset` , `Failed to connect` can be very generic error messages and may not be related to the configs...

File: src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java (hunk 1 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java:217]: SuccessFactorsUtil.isNotNullOrEmpty, this is being used in the whole project

File: src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java (hunk 2 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java:64]: added

File: src/main/java/io/cdap/plugin/successfactors/source/SuccessFactorsSource.java (hunk 3 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/successfactors/source/SuccessFactorsSource.java]: As tested the errors thrown in these scenarios are always same and to highlight the errors on the particular fields these keywords needs to be used.

File: src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java [hunk 4]
```diff
@@ -189,6 +197,39 @@ private Request buildRequest(URL endpoint, String mediaType) {
...
+   * Set to null or an empty string to bypass proxy configuration.
+   * @param proxyUsername  The username for proxy authentication. Set to null or an empty string if not required.
+   * @param proxyPassword  The password for proxy authentication. Set to null or an empty string if not required.
+   * @return An OkHttpClient configured with the specified proxy settings and authentication credentials.
+   */
+  private OkHttpClient buildConfiguredClient(String proxyUrl, String proxyUsername, String proxyPassword)
+    throws MalformedURLException, TransportException {
+    OkHttpClient.Builder builder = getConfiguredClient();
+
+    if (proxyUrl != null && !proxyUrl.isEmpty()) {
```
Comment (sahusanket) [inline src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java]: I meant here sorry 
use
`!Strings.isNullOrEmpty(proxyUrl)`
from
`import com.google.common.base.Strings;`

Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java]: done

File: src/main/java/io/cdap/plugin/successfactors/source/SuccessFactorsSource.java (hunk 3 above)
Comment (sahusanket) [reply src/main/java/io/cdap/plugin/successfactors/source/SuccessFactorsSource.java]: That might be true, 
but these key words may occur for other failures NOT related to proxy. 
Like network failure or something.

In such cases, these suggestion might mislead a user in debugging in wrong direction.

Comment (sahusanket) [reply src/main/java/io/cdap/plugin/successfactors/source/SuccessFactorsSource.java]: If there is no pre-defined error message for proxy related stuff, then Please surface the whole error. 
It's not correct to assume error message, as these can change in future unless it's mentioned by their official doc.

**Review by sahusanket — APPROVED** [review]

**Pull request merged**